### PR TITLE
Pull Request to Integrate LTSSM Log MRPC into switchtec_user callable functions

### PR DESCRIPTION
**Pull request by yungwesl**
This request covers the following changes:
- New code in main.c and switchtec.c to use the MRPC command LTSSM Monitor Command (LTMON) to dump link training information to a log (STDOUT)
- New structs defined in switchtec.h to support the output format
- Minor edit to temp_linux.py in the examples folder which recommends commenting out SET_MEAS as it is not required in the Gen4 switch but it is required in the Gen3 switch

Credits
- The original LTSSM Monitor code was written by Rakesh Vitta and checked into my repo (yungwesl) for convenience.  Although I maintain this code, the work was done entirely by Rakesh (so thank him and also blame him :))

Suggested Enhancements
- This code implements only the read function of the MRPC LTSSM Monitor Command
- There are some configurable items which can be used to enhance the behavior
-- Adding the ability to set the LTSSM mode to "nowrap" (however, this requires a link retrain following the config)

### CODE REVIEW

**Comment by PaulSelles** [comment]
Hi Wes,

The following changes should help clear up those build failures.  Looks like don't have permission to review the code directly within the PR so there are the diffs:

1. Remove link_rate_conversion declaration and changed from 'static const' to 'static'
```
diff --git a/cli/main.c b/cli/main.c
index 7c06186..cef8507 100644
--- a/cli/main.c
+++ b/cli/main.c
@@ -330,8 +330,6 @@ static char *pci_acs_to_string(char *buf, size_t buflen, int acs_ctrl,
        return buf;
 }

-char * link_rate_conversion(uint8_t link_rate);
-
 static const char* major_state_string[11] =
 {
         "Detect",
@@ -453,7 +451,7 @@ static const char* minor_state_string[11][12] =
     }
 };

-static const char * link_rate_conversion(uint8_t link_rate) {
+static char * link_rate_conversion(uint8_t link_rate) {
     switch(link_rate) {
          case 0:
               return "2.5G";
```

2. Move Gen 3 Switch check after argconfig_parse declaration and add print stderr
```
diff --git a/cli/main.c b/cli/main.c
index 0c77bc1..7097e77 100644
--- a/cli/main.c
+++ b/cli/main.c
@@ -482,11 +482,6 @@ static int ltssm_log(int argc, char **argv) {
                .phy_port = 0xff
        };

-       if (switchtec_is_gen3(cfg.dev)) {
-               fprintf (stderr, "This command is not supported on Switchtec Gen3\n");
-               return 0;
-       }
-
        ltssm_log_data ltssm_log_output_data[128];

        const struct argconfig_options opts[] = {
@@ -497,6 +492,11 @@ static int ltssm_log(int argc, char **argv) {

        argconfig_parse(argc, argv, CMD_DESC_LTSSM_LOG, opts, &cfg, sizeof(cfg));

+       if (switchtec_is_gen3(cfg.dev)) {
+               fprintf (stderr, "This command is not supported on Switchtec Gen3\n");
+               return 0;
+       }
+
        port = cfg.phy_port;
        ret = switchtec_ltssm_log(cfg.dev, port, &log_count, ltssm_log_output_data);
        printf("LTSSM Log for Physical Port %d (autowrap ON)\n\n",port);
```

File: inc/switchtec/switchtec.h [hunk 1]
```diff
@@ -355,8 +355,21 @@ switchtec_boot_phase(struct switchtec_dev *dev);
...
+typedef struct {
+     uint8_t  : 3;
+     uint8_t minor_state : 4;
+     uint8_t major_state : 4;
+     uint8_t  : 2;
+     uint8_t link_rate: 2;
+     uint32_t : 17;
+     uint32_t tstamp: 26;
+     uint8_t : 6;
+}__attribute__((packed)) ltssm_log_data;
```
Comment (kelvin-cao) [inline inc/switchtec/switchtec.h:369]: Bit-fields definition is vague as different compiler/machine may have different interpretations of the ordering of the bits. Normally we use shifting and masking to get proper value of some bits.

File: inc/switchtec/switchtec.h [hunk 2]
```diff
@@ -355,8 +355,21 @@ switchtec_boot_phase(struct switchtec_dev *dev);
...
+
+int switchtec_ltssm_log(struct switchtec_dev *dev, int port, int *log_count, ltssm_log_data *log_data);
 int switchtec_status(struct switchtec_dev *dev,
-		     struct switchtec_status **status);
+		struct switchtec_status **status);
```
Comment (kelvin-cao) [inline inc/switchtec/switchtec.h]: I guess it's unintended change.

File: cli/main.c [hunk 3]
```diff
@@ -334,17 +334,17 @@ char * link_rate_conversion(uint8_t link_rate);
 
 char* major_state_string[11] =
 {
-        "Detect ",
```
Comment (kelvin-cao) [inline cli/main.c]: Changes like this should be squashed, it's not good to add something in one commit, and correct it in another commit in the same PR.

File: cli/main.c [hunk 4]
```diff
@@ -464,11 +462,10 @@ char * link_rate_conversion(uint8_t link_rate) {
          case 2:
               return "8.0G";
          case 3:
-              return "16G";
```
Comment (kelvin-cao) [inline cli/main.c]: One commit for one thing, obviously this change has nothing to do with whitespace cleanup, and either be squashed if this was introduced in this PR, or use another commit to fix it.

**Comment by kelvin-cao** [comment]
> 
> 
> This request covers the following changes:
> 
>     * New code in main.c and switchtec.c to use the MRPC command LTSSM Monitor Command (LTMON) to dump link training information to a log (STDOUT)
> 
>     * New structs defined in switchtec.h to support the output format
> 
>     * Minor edit to temp_linux.py in the examples folder which recommends commenting out SET_MEAS as it is not required in the Gen4 switch but it is required in the Gen3 switch
> 
> 
> Credits
> 
>     * The original LTSSM Monitor code was written by Rakesh Vitta and checked into my repo (yungwesl) for convenience.  Although I maintain this code, the work was done entirely by Rakesh (so thank him and also blame him :))
> 
> 
> Suggested Enhancements
> 
>     * This code implements only the read function of the MRPC LTSSM Monitor Command
> 
>     * There are some configurable items which can be used to enhance the behavior
>       -- Adding the ability to set the LTSSM mode to "nowrap" (however, this requires a link retrain following the config)

Normally, one PR for one fix or one feature. The change to temp_linux.py seems irrelevant.

File: inc/switchtec/switchtec.h (hunk 2 above)
Comment (yungwesl) [reply inc/switchtec/switchtec.h]: Yes, I was hoping there was a way to reject a change so it simply reverts to the source as the editor may be introducing something

File: cli/main.c (hunk 4 above)
Comment (yungwesl) [reply cli/main.c]: Thanks, I will work to follow one-commit one change policy.

File: inc/switchtec/switchtec.h (hunk 1 above)
Comment (yungwesl) [reply inc/switchtec/switchtec.h:369]: Is there an example in the code base I can use as a template?

File: cli/main.c [hunk 5]
```diff
@@ -97,7 +97,7 @@ int switchtec_handler(const char *optarg, void *value_addr,
  * build is non-trivial. After evaluating the development effort
  * and the resources available, we have decided to remove Windows
  * build support from release roadmap.
- *
```
Comment (lsgunth) [inline cli/main.c:100]: Please don't change whitespace in big patchsets like this.

File: inc/switchtec/switchtec.h (hunk 1 above)
Comment (lsgunth) [reply inc/switchtec/switchtec.h:369]: Yes, never use bitfields for anything where you care where the bits actually are.

File: inc/switchtec/switchtec.h [hunk 6]
```diff
@@ -355,8 +355,21 @@ switchtec_boot_phase(struct switchtec_dev *dev);
...
+     uint8_t minor_state : 4;
+     uint8_t major_state : 4;
+     uint8_t  : 2;
+     uint8_t link_rate: 2;
+     uint32_t : 17;
+     uint32_t tstamp: 26;
+     uint8_t : 6;
+}__attribute__((packed)) ltssm_log_data;
+
+int switchtec_ltssm_log(struct switchtec_dev *dev, int port, int *log_count, ltssm_log_data *log_data);
```
Comment (lsgunth) [inline inc/switchtec/switchtec.h:371]: Why are we introducing a prototype for a function that does not exist yet? Introduce the prototype with the function.

File: inc/switchtec/switchtec.h (hunk 2 above)
Comment (lsgunth) [reply inc/switchtec/switchtec.h]: Yes, git is good at this, but you should be reviewing and cleaning up before creating a PR.

Before committing you can use `git add --patch` or `git commit --patch` to only select the changes you want to commit.

After committing, you may use `git reset --patch HEAD~1` to stage reverts for specific parts of the previous commit. (You'd then want to `git commit` and `git reset --hard`).

File: lib/switchtec.c [hunk 7]
```diff
@@ -398,6 +398,101 @@ static const char *lane_reversal_str(int link_up,
 	}
 }
 
+
+
+
+
```
Comment (lsgunth) [inline lib/switchtec.c:404]: Nit: don't add extra blank lines. Only one blank line between each function please.

File: lib/switchtec.c [hunk 8]
```diff
@@ -398,6 +398,101 @@ static const char *lane_reversal_str(int link_up,
...
+/**
+ * @brief Get the LTSSM log of a port on a switchtec device
+ * @param[in]  dev    Switchtec device handle
+ * @param[in]  port   Switchtec Port
+ * @param[out] log    A pointer to an array containing the log 
+ *
+ */
+
```
Comment (lsgunth) [inline lib/switchtec.c:412]: Nit: unnecessary blank line.

File: lib/switchtec.c [hunk 9]
```diff
@@ -398,6 +398,101 @@ static const char *lane_reversal_str(int link_up,
...
+int switchtec_ltssm_log(struct switchtec_dev *dev,
+		int port, int *log_count,
+		ltssm_log_data *ltssm_log_output_data) 
+{
+	int ret = 1;
```
Comment (lsgunth) [inline lib/switchtec.c:417]: Nit: prefer

```
	struct {
		uint8_t sub_cmd;
		uint8_t port;
		uint8_t freeze;
		uint8_t unused;
	} ltssm_freeze;
	int ret = 1;
```

(ie no blank line and largest declarations first).

File: lib/switchtec.c [hunk 10]
```diff
@@ -398,6 +398,101 @@ static const char *lane_reversal_str(int link_up,
...
+	ltssm_freeze.sub_cmd = 14;
+	ltssm_freeze.port = port;
+	ltssm_freeze.freeze = 1; //Freeze
+
+	ret = switchtec_cmd(dev, MRPC_DIAG_PORT_LTSSM_LOG, &ltssm_freeze, sizeof(ltssm_freeze), NULL, 0);
+
+	struct {
+		uint8_t sub_cmd;
+		uint8_t port;
+	}ltssm_status;
```
Comment (lsgunth) [inline lib/switchtec.c:435]: Don't declare things in the middle of a function. Prefer it all at the top, or in this case it looks like this function should be split into multiple static helpers -- at least one per switchtec_cmd().

File: lib/switchtec.c [hunk 11]
```diff
@@ -398,6 +398,101 @@ static const char *lane_reversal_str(int link_up,
...
+		uint8_t log_num;
+	}ltssm_status_output;
+
+	ltssm_status.sub_cmd = 13;
+	ltssm_status.port = port;
+	ret = switchtec_cmd(dev, MRPC_DIAG_PORT_LTSSM_LOG, &ltssm_status, sizeof(ltssm_status),
+			&ltssm_status_output, sizeof(ltssm_status_output)); 
+
+	*(log_count) = ltssm_status_output.log_num;
+//	printf("No.of logs for the port %d, is %d \n", port, ltssm_status_output.log_num);
```
Comment (lsgunth) [inline lib/switchtec.c:449]: Don't commit commented out code. And remove the extra white space.

File: cli/main.c [hunk 12]
```diff
@@ -330,6 +330,190 @@ static char *pci_acs_to_string(char *buf, size_t buflen, int acs_ctrl,
...
+        "INACTIVE",
+        "IDLE",
+        "TX_WAKE0",
+        "TX_WAKE1",
+        "EXIT",
+        "SPEED",
+    }
+};
+
+char * link_rate_conversion(uint8_t link_rate) {
```
Comment (lsgunth) [inline cli/main.c]: This function and everything above it should be static.

Also the style for functions is to have the brace on it's own line.

File: cli/main.c [hunk 13]
```diff
@@ -330,6 +330,190 @@ static char *pci_acs_to_string(char *buf, size_t buflen, int acs_ctrl,
...
+    }
+
+    return "2.5G";
+}
+
+#define CMD_DESC_LTSSM_LOG "Display LTMON log"
+static int ltssm_log(int argc, char **argv) {
+	int ret;
+	int port;
+        int log_count, i = 0;
```
Comment (lsgunth) [inline cli/main.c]: indent

File: cli/main.c [hunk 14]
```diff
@@ -330,6 +330,190 @@ static char *pci_acs_to_string(char *buf, size_t buflen, int acs_ctrl,
...
+
+	static struct {
+		struct switchtec_dev *dev;
+		int phy_port;
+	} cfg = {
+		.phy_port = 0xff
+	};
+
+
```
Comment (lsgunth) [inline cli/main.c:487]: More extra whitespace.

File: cli/main.c [hunk 15]
```diff
@@ -330,6 +330,190 @@ static char *pci_acs_to_string(char *buf, size_t buflen, int acs_ctrl,
...
+	port = cfg.phy_port;
+	ret = switchtec_ltssm_log(cfg.dev, port, &log_count, ltssm_log_output_data);
+
+	printf("Log Index  -  Timestamp  - Link Rate -  Major State  - Minor State \n");
+	for(i = 0; i < log_count ; i++) {
+		if(ltssm_log_output_data[i].major_state < 15) {
+		printf("%10d ", i);
+		printf("%12x ", ltssm_log_output_data[i].tstamp);
+		printf("%10s ", link_rate_conversion(ltssm_log_output_data[i].link_rate));
+		printf("%12s . ", major_state_string[ltssm_log_output_data[i].major_state]);
```
Comment (lsgunth) [inline cli/main.c]: What happens if the value is outside the array size. We should really check for that. We don't want to segfault if the firmware adds a new value.

File: cli/main.c [hunk 16]
```diff
@@ -330,6 +330,190 @@ static char *pci_acs_to_string(char *buf, size_t buflen, int acs_ctrl,
...
+        "SPEED",
+    }
+};
+
+char * link_rate_conversion(uint8_t link_rate) {
+
+
+    switch(link_rate) {
+         case 0:
+              return "2.5G";
```
Comment (lsgunth) [inline cli/main.c:459]: I'd find it acceptable to put each case on one line:

``` case 0: return "2.5G"; ```

File: cli/main.c [hunk 17]
```diff
@@ -547,6 +547,38 @@ static int bw(int argc, char **argv)
...
+		{"physical", 'f', "", CFG_NONNEGATIVE, &cfg.phy_port, required_argument,
+			"physical port ID"},
+		{NULL}};
+
+	argconfig_parse(argc, argv, CMD_DESC_LTMON, opts, &cfg, sizeof(cfg));
+
+	if (cfg.phy_port == 0xff)
+		printf("physical port: all\n");
+
+	ret = switchtec_get_port_ltmon_dmp(cfg.dev, cfg.phy_port);
```
Comment (lsgunth) [inline cli/main.c]: This doesn't exist yet, what's going on here?

File: cli/main.c [hunk 18]
```diff
@@ -731,38 +731,6 @@ static int bw(int argc, char **argv)
 	return 0;
 }
 
-#define CMD_DESC_LTMON "dump port ltssm state machine"
-static int ltmon(int argc, char **argv)
```
Comment (lsgunth) [inline cli/main.c]: Didn't we just add it in this PR? Reviewing code that gets removed or fixed later in the PR just wastes your reviewers time.

File: examples/temp_linux.py [hunk 19]
```diff
@@ -83,7 +83,8 @@ def echo_cmd(dev):
                            format(sub_cmd_in, sub_cmd_out))
 
 def die_temp(dev):
-    dev.cmd(MRPC_DIETEMP, struct.pack("<L", MRPC_DIETEMP_SET_MEAS))
+    # Comment out this next line for Gen4 switch
+    # dev.cmd(MRPC_DIETEMP, struct.pack("<L", MRPC_DIETEMP_SET_MEAS))
```
Comment (lsgunth) [inline examples/temp_linux.py:87]: First, this really doesn't belong in this PR which is supposed to be working on LTSSM stuff... This should be a separate PR.

Second, don't make broken code and don't leave code commented out. Probably best to find a way to check the generation and do the appropriate thing. Or maybe change the example to do a simpler MRPC command. (I only chose the temp one here because I thought it would be dead simple and stay that way).

File: cli/main.c [hunk 20]
```diff
@@ -476,11 +476,13 @@ char * link_rate_conversion(uint8_t link_rate) {
 static int ltssm_log(int argc, char **argv) {
 	int ret;
 	int port;
-        int log_count, i = 0;
+	int log_count, i = 0;
```
Comment (lsgunth) [inline cli/main.c:476]: Don't change code you've added in an earlier patch. Squash this with that patch.

File: cli/main.c [hunk 21]
```diff
@@ -501,12 +498,12 @@ static int ltssm_log(int argc, char **argv) {
...
 	ret = switchtec_ltssm_log(cfg.dev, port, &log_count, ltssm_log_output_data);
 	printf("LTSSM Log for Physical Port %d (autowrap ON)\n\n",port);
-	printf("Idx\tDeltaTime\tRate\tMajor_State\tMinor_State\n");
+	printf("Idx\tDeltaTime\tPCIeRate\tMajor_State\tMinor_State\n");
 	for(i = 0; i < log_count ; i++) {
 		if(ltssm_log_output_data[i].major_state < 15) {
 			printf("%3d\t", i);
 			printf("%09x\t", ltssm_log_output_data[i].tstamp);
-			printf("%4s\t", link_rate_conversion(ltssm_log_output_data[i].link_rate));
+			printf("%8s\t", link_rate_conversion(ltssm_log_output_data[i].link_rate));
```
Comment (lsgunth) [inline cli/main.c:511]: Again, squash the parts of this patch with the patches that are appropriate.

File: cli/main.c [hunk 22]
```diff
@@ -469,7 +469,7 @@ char * link_rate_conversion(uint8_t link_rate) {
     return "2.5G";
 }
 
-#define CMD_DESC_LTSSM_LOG "Display LTMON log"
+#define CMD_DESC_LTSSM_LOG "Display LTSSM log"
```
Comment (lsgunth) [inline cli/main.c:472]: Squash

File: inc/switchtec/switchtec.h [hunk 23]
```diff
@@ -372,7 +370,7 @@ typedef struct {
 
 int switchtec_ltssm_log(struct switchtec_dev *dev, int port, int *log_count, ltssm_log_data *log_data);
 int switchtec_status(struct switchtec_dev *dev,
-		struct switchtec_status **status);
+			      struct switchtec_status **status);
```
Comment (lsgunth) [inline inc/switchtec/switchtec.h:373]: This looks like it's being changed from correct to wrong... 

Don't make unnecessary whitespace changes.
